truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java

```diff
@@ -1085,4 +1088,32 @@ protected ObjectIVarSetNode createObjectIVarSetNode() {
 
     }
 
+    @CoreMethod(names = "store_exception", onSingleton = true, needsBlock = true)
+    public abstract static class StoreExceptionNode extends YieldingCoreMethodNode {
+
+        @Specialization
+        public Object executeWithProtect(VirtualFrame frame, DynamicObject block) {
+            try {
+                yield(frame, block);
+                return nil();
+            } catch (Throwable e) {
+                return e;
+            }
+        }
+    }
+
+    @CoreMethod(names = "raise_exception", onSingleton = true, required = 1)
+    public abstract static class RaiseExceptionNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        public Object executeWithProtect(Throwable e) {
+            if (e instanceof Error) {
+                throw (Error) e;
+            } else if (e instanceof RuntimeException) {
+                throw (RuntimeException) e;
+            }
+            throw new JavaException(e);
+        }
+    }
+
```

Comment:
So the idea of rb_protect and rb_jump_tag is to avoid unwinding the native stack in an uncontrolled manner. To do this we need to be able to run a piece of code and capture both its result (if it produces one), and any exception it generates. This is done by running the requested code in a block like this
```
e = store_exception { res = requested_function(...) }
```
leaving us with `e` containing any exception thrown from the block, and res containing the result of the function if it completed successfully. Since Ruby's API only allows us to use a number to indicate there was an exception we have to store it in a thread local array and provide a 1-based index into that array. Then once the native library has unwound its own stack by whatever method, and can allow the ruby error to propagate, it can call rb_jump_tag with the integer produced by rb_protect. rb_jump_tag then gets the exception out of the thread local and calls raise exception to throw it and allow normal error handling to continue.

Thinking about it I will rename `store_exception` to `capture_exception` since that feels like a better description of what it's doing.